File: ogn/client/ogn/Assets/script/shared/EventDispatcher.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


public class EventTarget
{
    public string name = "";
   	public EventDispatcher targetDispatcher = null;
	public EventRegister targetRegister = null;
}

public class EventRegister
{
    public static int s_id = 0;
    public int id = 0;
    public string name = "";
    public EventDispatcher.CallBack callback = null;
    public Object param = null;

    public EventRegister()
    {
        this.id = ++s_id;
    }
    public bool equalListener(string name, EventDispatcher.CallBack callback)
    {
        if (this.name == name && this.callback == callback)
		    return true;

	    return false;
    }
}

public class EventInvoke
{
    public virtual void Invoke(Object p1, Object p2)
    {

    }
}

public class EventCallBack<T1, T2> : EventInvoke
{
    public delegate int Callback(T1 lprama, T2 wprama);
    public Callback callback;
    public EventCallBack(Callback cb)
    {
        callback = cb;
    }
    public override void Invoke(Object t1, Object t2)
    {
        T1 p1 = (T1)t1;
        T2 p2 = (T2)t2;
        callback.Invoke(p1, p2);
    }
}

public class EventDispatcher
{
    public delegate int CallBack(EventTarget e);
    protected Dictionary<string, List<EventRegister>> listener = new Dictionary<string, List<EventRegister>>();
    protected Dictionary<int, List<EventInvoke>> listener_ = new Dictionary<int, List<EventInvoke>>();

    public EventDispatcher() {

    }

    public int addEventListener(string name, CallBack callback, Object param = null)
    {
        EventRegister lis = new EventRegister();
        lis.name = name;
	    lis.callback = callback;
	    lis.param = param;
        List<EventRegister> listeners = null;
        if (listener.ContainsKey(name)) {
            listeners = listener[name];
        }
        else {
            listeners = new List<EventRegister>();
            listener[name] = listeners;
        }

        listeners.Add(lis);

        return lis.id;
    }

    public int addEventListener(int id, EventInvoke callback)
    {
        List<EventInvoke> listeners = null;
        if (listener_.ContainsKey(id)) {
            listeners = listener_[id];
        }
        else
        {
            listeners = new List<EventInvoke>();
            listener_[id] = listeners;
        }
        listeners.Add(callback);
        return 0;
    }

    public int removeEventListener(string name, CallBack callback)
    {
        List<EventRegister> listeners = null;
        if (!listener.ContainsKey(name))
            return 0;
           
        listeners = listener[name];
        for (int i = 0; i < listeners.Count; ++i )
        {
            if (listeners[i].equalListener(name, callback))
            {
                listeners.RemoveAt(i);
                return 0;
            }
        }
        return 0;
    }
	public int dispatchEvent(EventTarget e)
    {
        List<EventRegister> listeners = null;
        if (!listener.ContainsKey(e.name))
            return 0;
           
        listeners = listener[e.name];
        for (int i = 0; i < listeners.Count; ++i)
        {
            e.targetDispatcher = this;
            e.targetRegister = listeners[i];
            listeners[i].callback.Invoke(e);
            //listeners[i].thisObject.(listeners[i].callback)(e);
        }
        return 0;
    }

    public int dispatchEvent(int id, Object lparam, Object wparam)
    {
        List<EventInvoke> listeners = null;
        if (!listener_.ContainsKey(id))
            return 0;

        listeners = listener_[id];
        for (int i = 0; i < listeners.Count; ++i) {
            listeners[i].Invoke(lparam, wparam);
        }
        return 0;
    }

}

File: ogn/src/Shared/core/core/EventDispatcher.cpp
#include "Shared.hpp"

IMPLEMENT_CLASS(EventDispatcher, Object)
EventDispatcher::EventDispatcher()
{
}

EventDispatcher::~EventDispatcher()
{
	std::map<std::string, std::vector<EventRegister*>>::iterator itr;
	for (itr = mMapListener.begin(); itr != mMapListener.end(); ++itr)
	{
		std::vector<EventRegister*>& listeners = itr->second;
		std::vector<EventRegister*>::iterator l_itr;
		for (l_itr = listeners.begin(); l_itr != listeners.end(); ++l_itr)
		{
			delete (*l_itr);
		}
		listeners.clear();
	}
	mMapListener.clear();
	
}

int32 EventDispatcher::addEventListener(const std::string& name, EventCallback callback, Object* thisObject, void* param /* = NULL */)
{
	EventRegister* lis = new EventRegister();

	lis->event = name;
	lis->thisObject = thisObject;
	lis->callback = callback;
	lis->param = param;

	std::map<std::string, std::vector<EventRegister*>>::iterator itr = this->mMapListener.find(name);
	std::vector<EventRegister*>* listeners = NULL;
	if (itr != this->mMapListener.end())
	{
		listeners = &(itr->second);
		listeners->push_back(lis);
	}
	else
	{
		std::vector<EventRegister*> ls;
		ls.push_back(lis);
		this->mMapListener.insert(std::make_pair(name, ls));
	}

	return lis->id;
}

int32 EventDispatcher::addEventListener(int32 id, EventCallbackProcess callback, Object* thisObject)
{
	auto itr = listener_.find(id);
	std::vector<std::pair<Object*, EventCallbackProcess>>* listeners = NULL;
	if (itr != this->listener_.end())
	{
		listeners = &(itr->second);
		listeners->push_back(std::make_pair(thisObject, callback));
	}
	else
	{
		std::vector<std::pair<Object*, EventCallbackProcess>> lst;
		lst.push_back(std::make_pair(thisObject, callback));
		listener_.insert(std::make_pair(id, lst));
	}
	return 0;
}

int32 EventDispatcher::addEventListener(int32 id, EventDispatcherCallback callback, Object* thisObject)
{
	auto itr = mMapEventDispatcher.find(id);
	if (itr != mMapEventDispatcher.end())
	{
		std::vector<std::pair<Object*, EventDispatcherCallback>>& lst = itr->second;
		lst.push_back(std::make_pair(thisObject, callback));
	}
	else
	{
		std::vector<std::pair<Object *, EventDispatcherCallback>> lst;
		lst.push_back(std::make_pair(thisObject, callback));
		mMapEventDispatcher.insert(std::make_pair(id, lst));
	}
	return 0;
}

int32 EventDispatcher::dispatch(int32 id, Parameter& par)
{
	auto itr = mMapEventDispatcher.find(id);
	if (itr == mMapEventDispatcher.end())
		return 0;

	auto lst = itr->second;

	for (auto pa : lst) {
		if (pa.first == NULL || pa.second == NULL) continue;
		(pa.first->*(pa.second))(par);
	}
	return 0;
}

int32 EventDispatcher::removeEventListener(const std::string& name, EventCallback callback, Object* thisObject)
{
	auto itr = this->mMapListener.find(name);
	std::vector<EventRegister*>* listeners = NULL;
	if (itr != this->mMapListener.end())
	{
		listeners = &(itr->second);

		for (auto l_itr = listeners->begin();
			l_itr != listeners->end();
			++l_itr)
		{
			EventRegister* lis = (*l_itr);
			if (lis->equalListener(name.c_str(), callback, thisObject))
			{
				listeners->erase(l_itr);
				if (listeners->size() <= 0)
					this->mMapListener.erase(itr);
				delete lis;
				return 0;

			}
		}
	}
	return 0;

}

int32 EventDispatcher::removeEventListener(int32 id, EventCallbackProcess callback, Object* thisObject)
{
	auto itr = listener_.find(id);
	if (itr != listener_.end())
	{
		auto& itrListeners = itr->second;
		for (auto itrList = itrListeners.begin(); itrList != itrListeners.end(); ++itrList)
		{
			if (callback == itrList->second && thisObject == itrList->first)
			{
				itrListeners.erase(itrList);
				break;
			}
		}

		if (itrListeners.size() <= 0)
			listener_.erase(itr);
	}

	return 0;
}

int32 EventDispatcher::dispatch(Event& event)
{
	int32 result = 0;
	auto itr = this->mMapListener.find(event.event);
	std::vector<EventRegister*>* listeners = NULL;
	if (itr != this->mMapListener.end())
	{
		listeners = &(itr->second);

		for (auto l_itr = listeners->begin();
			l_itr != listeners->end();
			++l_itr)
		{
			EventRegister* lis = (*l_itr);
			event.targetDispatcher = this;
			event.targetRegister = lis;
			event.param = lis->param;
			try
			{
				result = 1;
				if (lis->thisObject == NULL || lis->callback == NULL)
					continue;
				(lis->thisObject->*(lis->callback))(event);
			}
			catch (std::string e)
			{
				e = "Registration callback and the definition of the function is not set!";
				throw(e);
			}
		}
	}
	return result;
}

int32 EventDispatcher::dispatch(int32 id, void* lparam, void* wparam)
{
	int32 result = 0;
	auto itr = listener_.find(id);
	std::vector<std::pair<Object*, EventCallbackProcess>>* listeners = NULL;
	if (itr != this->listener_.end())
	{
		listeners = &(itr->second);

		for (std::vector<std::pair<Object*, EventCallbackProcess>>::iterator l_itr = listeners->begin();
			l_itr != listeners->end();
			++l_itr)
		{
			std::pair<Object*, EventCallbackProcess>& lis = (*l_itr);
			if (lis.second == NULL || lis.first == NULL) continue;

			(lis.first->*(lis.second))(lparam, wparam);
			result = 1;
		}
	}
	return result;
}
